### PR TITLE
UI5 Version validation for Adaptation project

### PR DESCRIPTION
**Pull request by GDamyanov**
Feat for [#1405](https://github.com/SAP/open-ux-tools/issues/1405)

Warning message when the ui5 version of the adp project is less than 1.71

### CODE REVIEW

File: packages/preview-middleware/package.json [hunk 1]
```diff
@@ -40,6 +40,7 @@
         "@sap-ux/logger": "workspace:*",
         "@sap-ux/adp-tooling": "workspace:*",
         "@sap-ux/control-property-editor-sources": "workspace:@sap-ux/control-property-editor@*",
+        "@sap-ux/ui5-config": "workspace:@sap-ux/ui5-config@*",
```
Comment (tobiasqueck) [inline packages/preview-middleware/package.json:43]: `"workspace:*"` is sufficient

**Review by tobiasqueck — CHANGES_REQUESTED** [review]
The implementation only covers the case with the Fiori tools middleware being the one defining the UI5 version, however, the project has other ways to set the version e.g. a different proxy middleware (there are some) or by defining it in the yaml config as local version.
Therefore, I don't think you can find a working implementation that just looks at the config, if you want to cover all cases, then you need to move your check to the client-side, during the initialization because only then you know the effective UI5 version.

File: packages/preview-middleware-client/src/adp/init.ts [hunk 2]
```diff
@@ -27,6 +29,15 @@ export default function (rta: RuntimeAuthoring) {
             }
         }
     });
+    
+    const ui5Version = (await VersionInfo.load()) as { version: string };
+    const ui5VersionValidationMessage = getUI5VersionValidationMessage(ui5Version.version);
+
+    if(ui5VersionValidationMessage) {
+        sendAction(showMessage('The current SAPUI5 version set for this Adaptation project is 1.70. The minimum version to use for SAPUI5 Adaptation Project and its SAPUI5 Visual Editor is 1.71'));
```
Comment (zdravko-georgiev) [inline packages/preview-middleware-client/src/adp/init.ts:37]: Please replace hardcoded test with the `ui5VersionValidationMessage` variable 😉

File: packages/ui5-proxy-middleware/src/ui5/middleware.ts [hunk 3]
```diff
@@ -86,6 +87,12 @@ module.exports = async ({ resources, options }: MiddlewareParameters<UI5ProxyCon
         logger.debug(error);
     }
 
+    if (await isUI5VersionRemoved(ui5Version)) {
+        throw new Error(
+            `You are trying to load the application preview with the non existing version ${ui5Version} of SAPUI5. Please check and correct the version.`
+        );
+    }
+
```
Comment (zdravko-georgiev) [inline packages/ui5-proxy-middleware/src/ui5/middleware.ts:95]: Please don't do any changes to the `ui5-proxy-middleware`. Any specific UI5 version handling needs to be done in a package that integrates the `ui5-proxy-middleware`. OR in this case either in `adp-tooling` or in the `preview-middleware-client`.

**Review by zdravko-georgiev — CHANGES_REQUESTED** [review]

File: packages/ui5-proxy-middleware/src/ui5/middleware.ts (hunk 3 above)
Comment (GDamyanov) [reply packages/ui5-proxy-middleware/src/ui5/middleware.ts]: Removed as discussed.

File: packages/preview-middleware-client/src/adp/init.ts (hunk 2 above)
Comment (GDamyanov) [reply packages/preview-middleware-client/src/adp/init.ts]: Fixed

File: packages/preview-middleware-client/src/adp/ui5-version-utils.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * @param minVersion to check
+ *
+ * @returns boolean
+ */
+export function isLowerThanMinimalUi5Version(version: string, minVersion: string): boolean {
+    if (version !== undefined && version !== '' && minVersion !== undefined && minVersion !== '') {
+        const minVersionParsed = parseUI5Version(minVersion);
+        const ui5VersionParsed = parseUI5Version(version);
+        if (
+            !isNaN(ui5VersionParsed.major) &&
```
Comment (nikmace) [inline packages/preview-middleware-client/src/adp/ui5-version-utils.ts:14]: For this big if statement could we split it more for readability? First, check if the version and minVersion are not empty. Second, check ui5VersionParsed.minor and ui5VersionParsed.major that they are valid numbers. Then check in the second if statement, which will be inside the first one, all the version stuff.

Something like this:

```typescript
if (version && minVersion) {
    const minVersionParsed = parseUI5Version(minVersion);
    const ui5VersionParsed = parseUI5Version(version);

    if (!isNaN(ui5VersionParsed.major) && !isNaN(ui5VersionParsed.minor)) {
        if (ui5VersionParsed.major < minVersionParsed.major) {
            return true;
        }
        if (ui5VersionParsed.major === minVersionParsed.major && ui5VersionParsed.minor < minVersionParsed.minor) {
            return true;
        }
    }
}
```

This also does not take into account the patch version (not sure if this is needed).

File: packages/preview-middleware-client/test/unit/adp/init.test.ts [hunk 5]
```diff
@@ -42,12 +38,56 @@ describe('adp', () => {
         });
     });
 
+    beforeEach(() => {
+        rtaMock.getDefaultPlugins
```
Comment (nikmace) [inline packages/preview-middleware-client/test/unit/adp/init.test.ts:42]: Why was this moved from the top, you mock the return value only once. You do not need to mock it again for each test.

File: packages/preview-middleware-client/test/unit/adp/init.test.ts [hunk 6]
```diff
@@ -42,12 +38,56 @@ describe('adp', () => {
...
+            controlId: 'v2flex::sap.suite.ui.generic.template.ListReport.view.ListReport'
+        };
+
+        await callBackFn({
+            type: '[ext] add-extension-point',
+            payload
+        });
 
-        init(rtaMock);
+        expect(executeSpy).toHaveBeenCalledWith(payload.controlId, 'CTX_ADDXML_AT_EXTENSIONPOINT');
```
Comment (nikmace) [inline packages/preview-middleware-client/test/unit/adp/init.test.ts:81]: This block of code is duplicated in two tests. It is testing the same thing two times, so maybe leave this block of code in the `init` test and remove it in the second test for the ui5 version notification.

File: packages/control-property-editor/src/slice.ts [hunk 7]
```diff
@@ -219,6 +222,9 @@ const slice = createSlice<SliceState, SliceCaseReducers<SliceState>, string>({
                     state.changes.controls[key] = control;
                 }
             })
+            .addMatcher(showMessage.match, (state, action: ReturnType<typeof showMessage>): void => {
+                state.showDialogMessage = action.payload;
```
Comment (nikmace) [inline packages/control-property-editor/src/slice.ts:226]: It would be good to rename this variable to dialogMessage. And also in the initialState.

So here when you match showMessage you set the state like this: 
```typescript
state.dialogMessage = action payload;
```

Also in `App.tsx` when retrieving from the store change showDialogMessage to dialogMessage.

**Review by nikmace — CHANGES_REQUESTED** [review]

File: packages/preview-middleware-client/test/unit/adp/init.test.ts (hunk 5 above)
Comment (GDamyanov) [reply packages/preview-middleware-client/test/unit/adp/init.test.ts:42]: This is done like that because for each call of init we should have these two specific mock return of getDefaultPlugins

File: packages/control-property-editor/src/App.tsx [hunk 8]
```diff
@@ -84,28 +86,42 @@ export default function App(appProps: AppProps): ReactElement {
...
             </section>
             <section className="app-panel app-panel-right">
                 <PropertiesPanel />
             </section>
+            <UIDialog
+                hidden={!showDialogMessage}
+                dialogContentProps={{
+                    title: t('TOOL_DISCLAIMER_TITLE'),
+                    subText: dialogMessage
+                }}></UIDialog>
```
[the remark this answers is not in the document]
Comment (GDamyanov) [reply packages/control-property-editor/src/App.tsx]: Done

File: packages/preview-middleware-client/test/unit/adp/init.test.ts [hunk 9]
```diff
@@ -42,12 +38,56 @@ describe('adp', () => {
...
+                }
+            })
+            .mockReturnValueOnce({});
+    });
+
+    afterEach(() => {
+        initOutlineSpy.mockClear();
+        addMenuItemSpy.mockClear();
+        setPluginsSpy.mockClear();
+        sendActionMock.mockClear();
```
[the remark this answers is not in the document]
Comment (GDamyanov) [reply packages/preview-middleware-client/test/unit/adp/init.test.ts]: Done

File: packages/preview-middleware-client/src/adp/init.ts [hunk 10]
```diff
@@ -27,6 +29,15 @@ export default function (rta: RuntimeAuthoring) {
             }
         }
     });
+    
+    const ui5Version = (await VersionInfo.load()) as { version: string };
+    const ui5VersionValidationMessage = getUI5VersionValidationMessage(ui5Version.version);
+
+    if(ui5VersionValidationMessage) {
```
[the remark this answers is not in the document]
Comment (GDamyanov) [reply packages/preview-middleware-client/src/adp/init.ts]: Accepted

File: packages/preview-middleware-client/src/adp/ui5-version-utils.ts (hunk 4 above)
Comment (GDamyanov) [reply packages/preview-middleware-client/src/adp/ui5-version-utils.ts:13]: Done

File: packages/preview-middleware-client/test/unit/adp/init.test.ts (hunk 6 above)
Comment (GDamyanov) [reply packages/preview-middleware-client/test/unit/adp/init.test.ts]: Done

File: packages/control-property-editor/src/slice.ts (hunk 7 above)
Comment (GDamyanov) [reply packages/control-property-editor/src/slice.ts]: Done

**Review by nikmace — APPROVED** [review]
Code looks good
Good test coverage
Manual testing ✅

File: .changeset/lazy-roses-drive.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+---
+'@sap-ux/preview-middleware': patch
+'@sap-ux/ui5-config': patch
+---
+
+Validate ui5 version defined in ui5-proxy-middleware for adp project
```
Comment (zdravko-georgiev) [inline .changeset/lazy-roses-drive.md:6]: Please fix the description. This is not true. The goal is to validate the UI5 version, regardless from where it is coming.

File: .changeset/lazy-roses-drive.md [hunk 12]
```diff
@@ -0,0 +1,6 @@
+---
+'@sap-ux/preview-middleware': patch
+'@sap-ux/ui5-config': patch
```
Comment (zdravko-georgiev) [inline .changeset/lazy-roses-drive.md:3]: We do not need to change `ui5-config`. Revert any any changes to `ui5-config` and remove this please.

File: packages/ui5-config/src/ui5config.ts
Comment (zdravko-georgiev) [inline packages/ui5-config/src/ui5config.ts:1]: Revert this change please

File: packages/ui5-proxy-middleware/tsconfig.json
Comment (zdravko-georgiev) [inline packages/ui5-proxy-middleware/tsconfig.json:1]: revert this change please

**Review by zdravko-georgiev — CHANGES_REQUESTED** [review]

**Review by zdravko-georgiev — CHANGES_REQUESTED** [review]

File: packages/preview-middleware/package.json (hunk 1 above)
Comment (zdravko-georgiev) [reply packages/preview-middleware/package.json]: Removed

File: packages/control-property-editor/src/App.tsx [hunk 13]
```diff
@@ -27,6 +28,12 @@ export interface AppProps {
  */
 export default function App(appProps: AppProps): ReactElement {
     const { previewUrl } = appProps;
+    const urlWithoutHash = previewUrl.split('#')[0];
+    const queryStr = urlWithoutHash.split('?')[1];
+    const urlParams = new URLSearchParams(queryStr);
+    const scenario = urlParams.get('sap-ui-scenario') as Scenario;
+    const isAdpProject = scenario === 'ADAPTATION_PROJECT';
```
Comment (zdravko-georgiev) [inline packages/control-property-editor/src/App.tsx:35]: Please use the `flexSettings` for reading the scenario https://github.com/SAP/open-ux-tools/blob/063ef51d1e2f0d906f4f295bc7776c32ab197a08/packages/preview-middleware-client/src/cpe/outline/index.ts#L17

**Review by zdravko-georgiev — CHANGES_REQUESTED** [review]

File: packages/control-property-editor/src/App.tsx (hunk 13 above)
Comment (GDamyanov) [reply packages/control-property-editor/src/App.tsx:35]: This is react app and we don't have access to rta.

Comment (zdravko-georgiev) [reply packages/control-property-editor/src/App.tsx:35]: @GDamyanov There is a communication channel between the react app and the preview where RTA is running. Please read the scenario from the RTA and then send it to the react app.

Comment (nikmace) [reply packages/control-property-editor/src/App.tsx:35]: > @GDamyanov There is a communication channel between the react app and the preview where RTA is running. Please read the scenario from the RTA and then send it to the react app.

This way there is a huge delay for the post messaging between the iframe and react application (different loading times). This is really visible for the user. The most efficient way is to pass scenario in `previewUrl` to the `start` method, which is the entry point of the react application. This way when the react application is loading we already know the scenario and we do not have a huge delay.

Comment (zdravko-georgiev) [reply packages/control-property-editor/src/App.tsx:35]: @nikmace Good point. If this is the case, then please add the `scenario` to the start options here https://github.com/SAP/open-ux-tools/blob/main/packages/control-property-editor/src/index.tsx#L28

Comment (zdravko-georgiev) [reply packages/control-property-editor/src/App.tsx:35]: @nikmace @GDamyanov It is possible to read the `scenario` from the store. It is already there. Please see here https://github.com/SAP/open-ux-tools/blob/e9f1ea561af740d92b54f43bb13114be6b634d6e/packages/control-property-editor/src/index.tsx#L28

Comment (nikmace) [reply packages/control-property-editor/src/App.tsx]: > @nikmace @GDamyanov It is possible to read the `scenario` from the store. It is already there. Please see here
> 
> https://github.com/SAP/open-ux-tools/blob/a9de3d97f3678e8d54fe5525530a84e382b3e273/packages/control-property-editor/src/App.tsx#L42

Where is this scenario coming from? From post-messaging? If yes, that is exactly what I was talking about we will now be getting a big delay before we actually know in what scenario we are. In App.tsx you should instead pass the scenario to the URL and dispatch it in App.tsx. And then you can get it from the store by either using a useSelector or directly using the variable which you retrieved from previewUrl.

The result of the last commit with reverted changes will be is that first the react application will be loaded, second, the iframe will start being rendered (and we will see white container), then we receive a post message from it with the scenario and the redux store updates and we stop showing the iframe and show a dialog message.  The time between the react application load time and the iframe is noticeable not to say the least.

Comment (zdravko-georgiev) [reply packages/control-property-editor/src/App.tsx]: @nikmace When I start the application the time between loading the react application and then the iframe is good. There is no significant delay.

File: packages/control-property-editor/src/App.tsx [hunk 14]
```diff
@@ -87,29 +89,45 @@ export default function App(appProps: AppProps): ReactElement {
         setWarningDialogVisibility(false);
     }
 
+    useEffect(() => {
+        if (dialogMessage && isAdpProject) {
+            setShouldShowDialogMessageForAdpProjects(true);
+        }
+    }, [dialogMessage]);
```
Comment (zdravko-georgiev) [inline packages/control-property-editor/src/App.tsx:96]: This is not working as expected. It is executed before actually the scenario is set to `ADAPTATION_PROJECT`. You can see it best when debugging. Please have a look.

**Review by zdravko-georgiev — CHANGES_REQUESTED** [review]

**Review by zdravko-georgiev — APPROVED** [review]
Code looks good
Patchset is OK
Review comments addressed
Very good test coverage
Manual testing ✅ 

Considering that the only available version on UI5 CDN that is below `1.71` is `1.38`, users should not get to see this pop-up too often.

**Review by nikmace — APPROVED** [review]
Code looks good
Review comments addressed
Good test coverage
Did not test manually

**Comment by nikmace** [comment]
Can someone from @SAP/ux-tools-control-property-editor do a review?

**Review by voicis — APPROVED** [review]
- CPE code looks good
- Good test coverage
- Manually tested CPE on FE projects.
